src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -56,6 +56,98 @@ DecisionTree<FitnessFunction,
   Train(data, labels, numClasses, minimumLeafSize);
 }
 
+//! Construct and train.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename ElemType,
+         bool NoRecursion>
+template<typename MatType>
+DecisionTree<FitnessFunction,
+             NumericSplitType,
+             CategoricalSplitType,
+             ElemType,
+             NoRecursion>::DecisionTree(MatType&& data,
+                                        const data::DatasetInfo& datasetInfo,
+                                        arma::Row<size_t>&& labels,
+                                        const size_t numClasses,
+                                        const size_t minimumLeafSize)
+{
+  // move the data to avoid copy
+  MatType tmpData = std::move(data);
+  arma::Row<size_t> tmpLabels = std::move(labels);
```

Comment:
I think const MatType& can't be replaced with MatType&& like following:
```cpp
  template<typename MatType>
  DecisionTree(MatType&& data,
               const arma::Row<size_t>& labels,
               const size_t numClasses,
               const size_t minimumLeafSize = 10){
MatType tmpData(std::forward<MatType>(data));
}
```
In this case, if I pass a lvalue mat<double>.
Then the MatType will be deduced as mat<double>&, not mat<double>.
When I tried this:
```cpp
template<typename T>
vector<int> F(T&& v) {
	T a = std::forward<T>(v);
	a.push_back(2);
	return a;
}
int main() {
	vector<int> x;
	x.push_back(33);
	vector<int> b = F(x);
	cout << x.size() << endl;
	cout << b.size() << endl;
	return 0;
}
```
It shows both x.size() and b.size() are 2.
Maybe I've not understood correctly? :)